### DISCUSSION TITLE
Pybricks C/C++ extension module compatibility

### DISCUSSION THREAD

**Discussion by Dhsjkd2** [pybricks/support#1565] [Q&A] [answered]
Hey I want to try split some more important processing like getting sensor reading and put it in a C or C++ library as I have heard thet are compatible with python. I'm not sure if it works with micropython. Please help

**Comment by dlech** [comment]
Can you explain more about what you hope to accomplish by doing this? For example, are you trying so solve a performance issue, are you just interested in learning how it works, etc.

**Reply by Dhsjkd2** [reply]
I am trying to put parts of the code like functions in a C library like thing to speed up the processing speed of those functions. These are then returned to the main program which uses the values to run. I am trying to use this for a RoboCup soccer robot.

**Reply by dlech** [reply]
Can you share the code you have that is too slow? There may be other ways to speed it up.

**Reply by Dhsjkd2** [reply]
Um the code is kinda confidential right now. But if you have any general suggestions on how to improve runtime performance that would be great.

**Reply by BertLindeman** [reply]
A cheep shot I asked Copilot (AI).

The response was:
Sure, here are some general tips to speed up your Pybricks Python code:

1. **Optimize Loops**: If you have a loop that's running many times, any improvements inside this loop will have a significant impact on your program's speed. For example, you could change the order of the bricks in the `for`-loop³.

2. **Avoid Blocking Operations**: If there's a part of your code that's waiting for something to happen (like a motor to reach a certain position), consider if there's a way to continue with other parts of your code instead of waiting. For instance, instead of pushing the element down for two seconds, you could stop sooner³.

3. **Use Efficient Methods**: Some methods might be more efficient than others. Experimenting with different methods can sometimes lead to performance improvements. For example, you could try experimenting with the `run_until_stalled` method³.

4. **Code Profiling**: Profiling your code can help you understand where the bottlenecks are. This involves measuring the time each part of your code takes to execute. Once you know which parts of your code are slow, you can focus your optimization efforts there.

Remember, the key to optimization is measuring. Don't just guess what might be slow - use profiling to know for sure. Also, make sure to test your code after each change to ensure it still works as expected. Happy coding! 😊

Source: Conversation with Bing, 3/25/2024

Some platitudes, but a reasonable starting point, I think.

[EDIT] I forgot to add the links provided:
([1](https://pybricks.com/projects/sets/boost/creative-toolbox/auto-builder/)) Precision building with Auto Builder - Pybricks: Python coding with LEGO. https://pybricks.com/projects/sets/boost/creative-toolbox/auto-builder/.
(2) Pybricks Documentation — pybricks v3.4.[1](https://pybricks.com/projects/sets/boost/creative-toolbox/auto-builder/) documentation. https://docs.pybricks.com/.
(3) Using Pybricks with Visual Studio Code - Python project for LEGO - Pybricks. https://pybricks.com/projects/tutorials/dev/tools/vscode/.

**Reply by dlech** [reply]
The biggest one is only use local variables and void any globals or `.` accessors of any attributes in any performance critical paths.

**Reply by Dhsjkd2** [reply]
Oh okay I will try and implement some of those basic optimisations to my code. Still regarding the C/C++ extension modules, I still kinda wanna know if they are compatible with Pybricks. Would be fun to know.

**Reply by Dhsjkd2** [reply]
And just to clarify about the performance issues, we have to use an infinite loop because our program runs in cycles and that is the most 'effective' way for a robocup soccer robot code. Some things like large loops cant really be avoided so I might have to optimize different aspects of the code. So the main logic kind of works like this:

One loop that checks if the Left hub button is pressed or the Right button (they start/stop the code)

Another loop that checks if it can recieve data from the small hub 

Another loop that is the decision maker based on whether it is in goalie or striker mode (main loop)

Inside the main loop it first reads all sensor values, depending on the values it then choses a direction to move and calls the .dc method to run the motors.

**Reply by laurensvalk** [reply]
If you have multiple loops nested inside each other and you have any wait statements in it, it can give the appearance of being slow. This can also be the case when waiting for buttons, sensors, and so on.

If that is case, it's perhaps not about code execution speed but more about how you design your code.

For some of these applications, multitasking can work quite well, if you aren't already using it. It won't magically make the actual execution faster, but it means one loop doesn't have to wait on the others. Each loop can separately wait on things like button presses while the other loops ("tasks") just keep running.

**Reply by jaguilar** [reply, marked as answer]
@Dhsjkd2 Regarding your question about C extensions: to my understanding, pybricks can use the same extension methods as vanilla micropython. For pybricks 3, see https://docs.micropython.org/en/latest/develop/natmod.html (note that this does not work on the movehub as of the last time I checked). And for pybricks 2 on ev3, you'd use the ffi module, which I built an example for here: https://github.com/jaguilar/pybricks-ev3-ffi-demo and used in a project here: https://github.com/jaguilar/gyroboy/blob/main/pid.c. There is also the Viper code generator (https://docs.micropython.org/en/v1.9.3/pyboard/reference/speed_python.html#the-viper-code-emitter).

One common thing about all these native solutions is that they are very much a pain to use, in different ways. Viper is quite fiddly. Getting a cross-compilation environment working is tricky. So if you can make the Python fast enough, it's best to just do that. However, if it's borderline, at least by my measurements, C is indeed much faster than even the fastest Python code.

**Reply by Dhsjkd2** [reply]
@jaguilar hey thanks for that information. Does pybricks have specific API for C extensions?

**Reply by jaguilar** [reply]
@Dhsjkd2 not that I'm aware of.
